### PR TITLE
fix: block complete messenger may execute in main thread

### LINKED ISSUE #1069: blockComplete 在主线程回调
#### Before Issue

1. Please search on the [Issues](https://github.com/lingochamp/FileDownloader/issues)
2. Please search on the [wiki](https://github.com/lingochamp/FileDownloader/wiki)
3. Please set `FileDownloadLog.NEED_LOG=true` and review the Logcat output from main process and `:filedownloader` process ( pay attention to Warn and Error level logcat)

#### Issue

1. What problem do you get?
2. Which version of FileDownloader are you using when you produce such problem?
3. How to reproduce such problem?
4. Do you set `FileDownloadLog.NEED_LOG=true`?
5. Could you please reproduce this problem and provide all main process and `:filedownloader` process logcat 
6. Can you fix it by yourself and request PR, if not, what's problem do you get when you try to fix it

>P.S. If you don't know how to get `:filedownloader` process, it's recommended to using `pidcat` to just filter all your application logcat, or define `process.non-separate=true` on [filedownloader.properties](https://github.com/lingochamp/FileDownloader/wiki/filedownloader.properties)

---

blockComplete 按照注释应该是在非UI线程回调，但是在使用中发现并不是这样的
打log看了下，原因如下：
1. FileDownloadList 在文件下载完成后，remove方法会回调，switch 走FileDownloadStatus.completed 分支，然后在 FileDownloadMessenger 中的parcelQueue队列后面加入BlockCompleteMessageImpl
2. 但是这时候，parcelQueue可能不为空，里面还有没处理完的 ProgressMessageSnapshot
3. FileDownloadMessageStation 的 requestEnqueue 判断是否走blockCompletedPool 的逻辑是parcelQueue头，是否是 BlockCompleteMessageImpl。但是这时候 parcelQueue 头是还没处理完的 ProgressMessageSnapshot，BlockCompleteMessageImpl其实在队列尾部。这样就造成了这个不会走blockCompletedPool，而是走UI线程。
4. 如上所述，这个设计本身就隐形的要求 BlockCompleteMessageImpl 入队列时，队列为空？如果不为空就不走UI线程？ 为什么不能在 BlockCompleteMessageImpl 入队列时，清空一下队列？

### PR DESCRIPTION
**Pull request by rantianhua**
Closes #1069

### CODE REVIEW

File: library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java [hunk 1]
```diff
@@ -102,6 +96,25 @@ private void handoverInUIThread(IFileDownloadMessenger messenger) {
         handler.sendMessage(handler.obtainMessage(HANDOVER_A_MESSENGER, messenger));
     }
 
+    /**
```
Comment (Jacksgong) [inline library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java]: what about `interceptForBlockCompleteMessage`? I think logic is not changed?

File: library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java [hunk 2]
```diff
@@ -166,6 +179,9 @@ public boolean handleMessage(Message msg) {
         private void dispose(final ArrayList<IFileDownloadMessenger> disposingList) {
             // dispose Sub-package-size each time.
             for (IFileDownloadMessenger iFileDownloadMessenger : disposingList) {
+                if (handleBlockCompleteMessage(iFileDownloadMessenger)) {
```
Comment (Jacksgong) [inline library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java]: `blockCompleteMessage` handled on the `dispose` area is not make sense, another better way?  I think all message need filter with `requestEnqueue`, so why it handles on the `requestEnqueue` is not enough?

File: library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java (hunk 1 above)
Comment (rantianhua) [reply library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java]: Ok, this name is better.

File: library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java (hunk 2 above)
Comment (rantianhua) [reply library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java]: Yes, `blockCompleteMessage` should be handled on `requestEnqueue` method:
```java
    void requestEnqueue(final IFileDownloadMessenger messenger,
                        @SuppressWarnings("SameParameterValue") boolean immediately) {

        if (messenger.handoverDirectly()) {
            messenger.handoverMessage();
            return;
        }

        if (messenger.isBlockingCompleted()) {
            blockCompletedPool.execute(new Runnable() {
                @Override
                public void run() {
                    messenger.handoverMessage();
                }
            });
            return;
        }

        if (!isIntervalValid()) {
            ....
        }

        if (!isIntervalValid() || immediately) {
            ....
        }

        // enqueue.
        enqueue(messenger);
    }
```
But we can see, it depends on that `messenger.isBlockingCompleted()` return true, however, if `FileDownloadMessenger#parcelQueue` has unhandled progress message(it's possible if the consumer is slower than producer.),  `messenger.isBlockingCompleted()` will return false and the `blockCompleteMessage` will be added to `waitingQueue` and be handled on the `dispose` method at later.

File: library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java (hunk 1 above)
Comment (rantianhua) [reply library/src/main/java/com/liulishuo/filedownloader/FileDownloadMessageStation.java]: Updated.

**Pull request merged**